src/docfx/build/Build.cs

```diff
@@ -77,56 +88,67 @@ bool ShouldBuildFile(Document file)
         {
             try
             {
+                var model = (object)null;
+                var dependencies = DependencyMap.Empty;
+                var errors = Enumerable.Empty<Error>();
+
                 switch (file.ContentType)
                 {
                     case ContentType.Resource:
-                        return await BuildResource(context, file);
+                        BuildResource(context, file);
+                        return dependencies;
                     case ContentType.Markdown:
-                        return await BuildMarkdown.Build(context, file, tocMap, contribution, buildChild);
+                        (errors, model, dependencies) = await BuildMarkdown.Build(file, tocMap, contribution, buildChild);
+                        break;
                     case ContentType.SchemaDocument:
-                        return await BuildSchemaDocument.Build();
+                        (errors, model, dependencies) = await BuildSchemaDocument.Build();
+                        break;
                     case ContentType.TableOfContents:
-                        return await BuildTableOfContents.Build(context, file, tocMap, buildChild);
+                        (errors, model, dependencies) = BuildTableOfContents.Build(file, tocMap, buildChild);
+                        break;
                     case ContentType.Redirection:
-                        return await BuildRedirectionItem(context, file);
-                    default:
-                        return DependencyMap.Empty;
+                        model = BuildRedirectionItem(file);
+                        break;
+                }
+
+                var hasErrors = context.Report(file.ToString(), errors);
+                if (model != null && !hasErrors)
+                {
+                    context.WriteJson(model, file.OutputPath);
+                    return dependencies;
                 }
+
+                return null;
             }
             catch (DocfxException ex)
             {
                 context.Report(file.ToString(), ex.Error);
-                return DependencyMap.Empty;
+                return null;
```

Comment:
>return null; [](start = 16, length = 12)

here, you use `null` to represent a special status of `BuildFile` function, please add some return value description incase someone edit this in the future

Reply:
right, i should use a tuple instead.

Reply:
that would be better

---
In reply to: [202225993](https://github.com/dotnet/docfx/pull/3064#discussion_r202225993) [](ancestors = 202225993)